engine/src/main/java/org/pentaho/di/trans/ael/websocket/SessionConfigurator.java

```diff
@@ -0,0 +1,223 @@
+  private static final String ERROR_MSG = "Error starting websocket session";
+
+  private static final String USE_KEY_TAB_OPT = "useKeyTab";
+  private static final String KEY_TAB_OPT = "keyTab";
+  private static final String PRINCIPAL_OPT = "principal";
+  private static final String STORE_KEY_OPT = "storeKey";
+  private static final String USE_TICKET_CACHE_OPT = "useTicketCache";
+  private static final String DO_NOT_PROMPT_OPT = "doNotPrompt";
+  private static final String IS_INITIATOR_OPT = "isInitiator";
+
+  private boolean withAuth = false;
+  private String principal = null;
+  private String keytab = null;
+  private URI url;
+
+  private LoginContext loginContext;
+
+  private static final Credentials credentials = new NullCredentials();
+  private static final Header AUTHENTICATE_HEADER = new Header() {
+    @Override public String getName() {
+      return WWW_AUTHENTICATE;
+    }
+
+    @Override public String getValue() {
+      return NEGOTIATE;
+    }
+
+    @Override public HeaderElement[] getElements() throws ParseException {
+      return new HeaderElement[ 0 ];
+    }
+  };
+
+  public SessionConfigurator( URI url, String keytab, String principal ) {
+    this.url = url;
+    this.withAuth = !Strings.isNullOrEmpty( principal );
+    this.principal = principal;
+    this.keytab = keytab;
+  }
+
+  @Override
+  public void beforeRequest( Map<String, List<String>> headers ) {
+    if ( withAuth ) {
+      Header authenticationHeader = getAuthenticationHeader( url );
+      if ( authenticationHeader != null ) {
+        headers
+          .put( authenticationHeader.getName(), Collections.singletonList( authenticationHeader.getValue() ) );
+      }
+    }
+  }
+
+  @Override
+  public void afterResponse( HandshakeResponse hr ) {
+    try {
+      if ( loginContext != null ) {
+        loginContext.logout();
+      }
+    } catch ( LoginException e ) {
+      e.printStackTrace();
+      //work is done just ignore
+    }
+  }
+
+  private Header getAuthenticationHeader( URI uri ) throws RuntimeException {
+    try {
+      ClientLoginConfig loginConfig =
+        new ClientLoginConfig( this.keytab, this.principal );
+
+      Subject serviceSubject = getServiceSubject( loginConfig );
+      return Subject.doAs( serviceSubject, new PrivilegedAction<Header>() {
+        public Header run() {
+          // First try without stripping the port
+          RuntimeException saveFirstException;
+          try {
+            return spnegoAuthenticate( false, uri );
+          } catch ( Exception e ) {
+            saveFirstException = new RuntimeException( e );
+          }
+          // if fails let's try stripping the port
+          try {
+            return spnegoAuthenticate( true, uri );
+          } catch ( Exception e ) {
+            //let's send the first exception
+            throw saveFirstException;
+          }
+        }
+      } );
+    } catch ( RuntimeException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)